### ISSUE TITLE
timeout http-keep-alive doesn't work for backend servers

### ISSUE THREAD

**Issue by sergiimk** [haproxy/haproxy#1236]
## Detailed description of the problem

Specifying `timeout http-keep-alive 10s` in the `backend` section doesn't seem to have any effect. Haproxy keeps idle server connections open for very long periods of time.

This is very problematic in the following scenario:
* haproxy runs as an ingress controller in kubernetes, fronting traffic to multiple application pods
* application pods are being restarted
* graceful shutdown procedure of the apps
  * starts attaching `Connection: close` to all responses to make haproxy close connection
  * expects all idle connections to be closed in `timeout http-keep-alive 10s`
* haproxy does not obey the timeout
* application has to forcefully close the TCP connection after 20s
* this creates a race condition where connection being closed may still be used by haproxy for a new request
* this results in few hundreds of errors (HTTP 502) every time we re-deploy application pods

## Expected behavior

When `timeout http-keep-alive 10s` is set in the `backend` section haproxy will close all idle connections after 10s, allowing us to re-deploy our application pods with no traffic impact.

## Steps to reproduce the behavior

1. Run any HTTP app of your choice that supports keep-alive connections
2. Run haproxy with provided config (I'm using latest dev docker image)
```sh
docker run \
  -it --rm -p 8080:80 \
  -v `pwd`/haproxy.cfg:/usr/local/etc/haproxy/haproxy.cfg \
  --name haproxy --sysctl net.ipv4.ip_unprivileged_port_start=0 \
  haproxy:2.4-dev16 -dV -f /usr/local/etc/haproxy/haproxy.cfg
```
3. Use `curl` to make HTTP request to your app via haproxy
4. Observe that TCP connection is still open after 10s (via application logs or `netstat`)

## Do you have any idea what may have caused this?

Recent issue #984 seems related. I believe the fix is addressing the similar situation on the `frontend` level, but not the backed connections.

## Do you have an idea how to solve the issue?

Code doesn't seem to be handling `timeout.httpka` on the backend at all :(

## What is your configuration?

```sh
global
    daemon

defaults
    log global
    option dontlognull
    timeout client 1s
    timeout connect 1s
    # Always store idle connections to the backend servers and allow any
    # request to use existing idle connections
    http-reuse always

backend app_be
    mode http
    balance roundrobin
    timeout connect 100ms
    timeout http-keep-alive 10s
    timeout http-request 1s
    timeout server 1s
    server srv001 172.17.0.2:10001 weight 1 maxconn 100 check inter 2s

frontend http_fe
    mode http
    bind :80
    timeout client 1s
    default_backend app_be
```

## Output of `haproxy -vv` and `uname -a`

```sh
HA-Proxy version 2.4-dev16-86512dd 2021/04/09 - https://haproxy.org/
Status: development branch - not safe for use in production.
Known bugs: https://github.com/haproxy/haproxy/issues?q=is:issue+is:open
Running on: Linux 5.10.23-1-MANJARO #1 SMP PREEMPT Thu Mar 11 18:47:18 UTC 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_PROMEX=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=12).
Built with OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
Running on OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.32 2018-09-10
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 8.3.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace
```

```
Linux 42a7806de0a3 5.10.23-1-MANJARO #1 SMP PREEMPT Thu Mar 11 18:47:18 UTC 2021 x86_64 GNU/Linux
```

## If HAProxy crashed: Last outputs and backtraces

No crash

## Additional information (if helpful)

Same invalid behavior is observed on version `2.0.14`

**Comment by capflam** [comment]
`http-keep-alive` timeout is applied on the client connection, not the server one. Setting it in a backend section does not mean it is used on server connection. It it the maximum time to wait for the next request, thus on client side.
Idle connections are closed by a periodic task, triggered every `pool-purge-delay` interval (5s by default). Every time the task is executed, half of idle connections are closed. If the server closes an idle connection, it is removed from the idle list (except there is a bug). If you get 502 errors, It may be a timing issue, when an idle connection is reused at the time the server closes it.

But I'm curious, you described the way you shut down the old application to spawn the new one, but not how you handle the switch. You have only one server in your configuration. Thus if the procedure is to stop a pod to spawn a new one on the same ip/port, of course you may have 502 errors.

At first glance, it is not a bug but a configuration issue.

**Comment by sergiimk** [comment]
Thanks for looking into this, @capflam.

> But I'm curious, you described the way you shut down the old application to spawn the new one, but not how you handle the switch.

We're using [haproxy-ingress controller](https://github.com/jcmoraisjr/haproxy-ingress). During the rollout:
* Few new pods are created
* Controller adds them as servers into haproxy (dynamically via control socket or by reloading if there are no free slots)
* Corresponding amount of old pods is signaled to shutdown
* Old pods are disabled in haproxy by the controller (dynamically via socket)
* Process continues until all pods are replaced

So in this process we never have a situation when there is no servers to handle the traffic, and we never reuse ip/port. 

> If you get 502 errors, It may be a timing issue, when an idle connection is reused at the time the server closes it.

It is. Wireshark shows haproxy reusing an idle connection of the server that is shutting down. At the same time server gives up waiting for haproxy to close connection and terminates it, so haproxy gets an RST in response and propagates it to the client as HTTP 502. So I'm looking for a way to avoid this race condition and make haproxy always close connections on its side before they time-out on the server.

> Idle connections are closed by a periodic task, triggered every `pool-purge-delay` interval (5s by default). Every time the task is executed, half of idle connections are closed.

I tried adding `default-server pool-purge-delay 1s` and making one request - just like before the connection idles until server gives up on it after >10s. The *"half of idle connections are closed"* is very confusing to me ... what is considered idle? Will connection that just served a request 1ms ago still be considered idle by this purge logic? Why only half - some sort of heuristic? Why not close connections that were idle over certain period of time?

It seems that setting low `pool-purge-delay` may help us to reduce the number of errors but not completely eliminate them, yet at the same time it will also significantly reduce the utility of keep-alive since we'll bee closing connections too often even during normal operations.

Please let me know if I'm missing something.

**Comment by capflam** [comment]
Ok thanks for clarification. If the server connection is closed when HAProxy tries to reuse it, there is nothing to do except redispatching the request on retry. But it means the connection was not reused for 20 seconds. It means you have a very low traffic. Or I missed something.

About the purge mechanism, it is indeed an heuristic to maximize the reuse rate. It is also a good way to not wake up the purge task too often. The algorithm is in fact a bit more complex.  We remove half the difference between the total of established connections (used or idle) and the max number of used connections. However, you must not evaluate it on a very low traffic because on 2.2 and upper, the health-check on a server will count for one in used connections. Thus, it is possible to always keep one idle connection.

BTW, I'm a bit puzzled, because you said the server is disabled on the CLI, thus it must not be elected to process new requests. And because health-checks are disabled for such servers, idle connections should be purged.

**Comment by sergiimk** [comment]
> But it means the connection was not reused for 20 seconds. It means you have a very low traffic. Or I missed something.

We have moderate traffic when reproducing this issue under a load test. There's probably a lot of factors that can result in one idle connection not being utilized much longer than others (which haproxy thread handles it, load balancing algo etc.).

I still think it's a bit weird not to have a mechanism to reliably avoid such race conditions by having haproxy-side timeout shorter than the server-side.

> BTW, I'm a bit puzzled, because you said the server is disabled on the CLI, thus it must not be elected to process new requests. And because health-checks are disabled for such servers, idle connections should be purged.

This is a great question. I tried to see exactly what state transitions are happening when pod is replaced, and saw a very interesting picture:
* Run 1 haproxy instance and 1 pod
* Do few request via `curl` to make haproxy open some Keep-Alive connections to the server
* Trigger redeployment
* New pod goes up and is added by the ingress controller via socket - it goes `UP` almost instantly
* Old pod enters `Terminating` status in k8s
* Immediately marked as `MAINT` in haproxy by the controller

I see haproxy (note we still run version `2.0.14`) not purging the idle connections when server goes into `MAINT`.

Even on the stats screen you can see that connections are still idle:
![image](https://user-images.githubusercontent.com/204914/116611439-37b01680-a8eb-11eb-9e1b-3d08d580d954.png)

Only after server's graceful stop routine times out and force-closes the connections this counter goes to zero (means it's not stale because of `MAINT` state).

@capflam, is it expected that haproxy purges idle connections during the state transition into `MAINT`, or only upon next `pool-purge-delay` "tick"?

What's worse is that in multi-pod scenario I observed that right after one server went into `MAINT` it was soon re-used by the controller for a newly-started pod (given a different IP and pulled out of `MAINT` via control socket) and went into `UP` state while still showing **non-zero idle connection count**. This could explain why we see haproxy sending requests to the pods that should already be considered out of service.

I will now try to reproduce this situation on `2.4-dev`.

**Comment by sergiimk** [comment]
I have just confirmed that putting a server into "maint" mode in haproxy `2.0.14` does not purge idle connections, and neither does assigning a new IP. If server quicky transitions `ready -> maint -> new IP -> ready` new requests will reuse the old idle connections that go to server that no longer exist.

I have also tested the same with `2.4-dev` and the problem seems fixed. This was most likely the source of all our 502 errors, so we'll have to upgrade.

It still seems possible to get into a connection close race-condition under normal operations (without re-deploying servers), so it would be nice if `timeout http-keep-alive 10s` in the backend section actually applied to server connections. But a combination of a low-ish `pool-purge-delay` and high-ish application-side keep-alive timeout should help there, so I'm ready to move on.

Thanks a lot for your help @capflam and shedding light on the logic behind this!

**Comment by capflam** [comment]
Indeed, there is a difference in the way idle connections are handled between 2.0 and >=2.2. Especially, there is a fix pushed to 2.2 for this specific use-case (6318d33ce62). It was first backported to the 2.0 and reverted because there are too much differences between both versions. So upgrading at least to 2.2 is a good idea here.

About the normal case, you should be prepared to have some connection closures during reuse. It may happen regardless any server keep-alive timeout. To avoid errors when this happens, you may configure L7 retries for empty responses (via `retry-on  conn-failure empty-response`) and redispatch the request on another server (using `option redispatch`). But if it happens for running servers only, configuring the redispatch is not necessary.

**Comment by capflam** [comment]
Note that another solution is to use the `safe` strategy to reuse connections.
